dht-cache/src/cache.rs

```diff
@@ -0,0 +1,462 @@
+    };
+
+    let stream = UnboundedReceiverStream::new(r);
+
+    let local_read = local.clone();
+    let cmd_update = cmd.clone();
+    let peer_id = local_peer_id.clone();
+
+    tokio::task::spawn(async move {
+        let mut interval = time::interval(Duration::from_millis(resend_interval.max(100)));
+        while !cmd_update.is_closed() {
+            interval.tick().await;
+            let hash = local_read.get_hash().await;
+            let m = DomoCacheStateMessage {
+                peer_id: peer_id.clone(),
+                cache_hash: hash,
+                publication_timestamp: utils::get_epoch_ms(),
+            };
+
+            if cmd_update
+                .send(Command::Config(serde_json::to_value(&m).unwrap()))
+                .is_err()
+            {
+                break;
+            }
+        }
+    });
+
+    // TODO: refactor once async closures are stable
+    let events = stream.filter_map(move |ev| {
+        let local_write = local.clone();
+        let peers_state = peers_state.clone();
+        let peer_id = local_peer_id.clone();
+        let cmd = cmd.clone();
+        async move {
+            match ev {
+                Event::Config(cfg) => {
+                    let m: DomoCacheStateMessage = serde_json::from_str(&cfg).unwrap();
+
+                    let hash = local_write.get_hash().await;
+
+                    let republish = {
+                        let mut peers_state = peers_state.write().await;
+
+                        // update the peers_caches_state
+                        peers_state.insert(m);
+
+                        // check for desync
+                        let sync_info = peers_state.is_synchronized(&peer_id, hash);
+
+                        log::debug!("local {peer_id:?} {sync_info:?}  -> {peers_state:#?}");
+
+                        if let CacheState::Desynced { is_leader } = sync_info {
+                            is_leader
+                                && utils::get_epoch_ms() - peers_state.last_repub_timestamp
+                                    >= peers_state.repub_interval
+                        } else {
+                            false
+                        }
+                    };
+
+                    // republish the local cache if needed
+                    if republish {
+                        local_write
+                            .read_owned()
+                            .await
+                            .mem
+                            .values()
+                            .flat_map(|topic| topic.values())
+                            .for_each(|elem| {
+                                let mut elem = elem.to_owned();
+                                log::debug!("resending {}", elem.topic_uuid);
+                                elem.republication_timestamp = utils::get_epoch_ms();
+                                cmd.send(Command::Publish(serde_json::to_value(&elem).unwrap()))
+                                    .unwrap();
```

Comment:
```suggestion
                               
                                // This cannot fail because `cmd` is the sender part of the
                                // `stream` we are currently reading. In practice, we are
                                // queueing the commands in order to read them later.
                                cmd.send(Command::Publish(serde_json::to_value(&elem).unwrap()))
                                    .unwrap();
```